### PR TITLE
fix: add lock for kubernetesOpenAPIVersion

### PR DESCRIPTION
**Pull request by pmalek**
Ideally we'd introduce a fully fledged fix that avoids usage of globals but that might be more involved hence this simple fix.

Partially fixes: #4824.

### CODE REVIEW

**Comment by pmalek** [comment]
> Hi @pmalek. Thanks for your contribution!
> 
> I think we need to check in the test whether #4824 is fixed. Could you add test cases to check that problem is fixed?

Hi @koba1t 👋 

What we're trying to test here is a data race which is tricky to deterministically ensure in tests.

I can submit a test that would spin 2 (or more?) goroutines which would call `SetSchema`/`ResetOpenAPI` for a while and verify that it didn't `panic` but in order for this to make sense we'd need to add a `-race` flag to `go test` command invocation. I'd assume this is the relevant one: https://github.com/kubernetes-sigs/kustomize/blob/4fc02497ae7de4b1f7e31656e13720d4e719dcc0/kyaml/Makefile#L36

Can you give some guidance on this?

**Comment by koba1t** [comment]
Hi @pmalek!

That is the right line to edit the test flag.

> Can you give some guidance on this?

I think we just edit the `kyaml/Makefile` file.

**Comment by pmalek** [comment]
I've added a test that tests this (to an extent possible). I'm not particularly fond of this solution but hey! it helps with the data race problem.

**Review by KnVerey — COMMENTED** [review]
/approve

Thanks for adding a test! I confirmed that it fails pretty reliably on master; it just needs a real name, please. :)

File: kyaml/openapi/openapi_test.go [hunk 1]
```diff
@@ -325,3 +327,69 @@ func TestIsNamespaceScoped_custom(t *testing.T) {
 	assert.True(t, isFound)
 	assert.True(t, isNamespaceable)
 }
+
+func TestX(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pmalek) [reply kyaml/openapi/openapi_test.go]: 😬 My bad. PTAL now or suggest a different name please.

**Comment by KnVerey** [comment]
The test failures are unrelated and fixed by https://github.com/kubernetes-sigs/kustomize/pull/4995

**Comment by koba1t** [comment]
/lgtm

Could you try to `rebase master` to fix test failures?

**Comment by pmalek** [comment]
> /lgtm
> 
> Could you try to `rebase master` to fix test failures?

Done

**Comment by KnVerey** [comment]
I'm so sorry, but an external change caused another failure on master that was fixed in #5004 . Can you please rebase one last time? I swear this isn't normal. 😅

**Comment by pmalek** [comment]
> I'm so sorry, but an external change caused another failure on master that was fixed in #5004 . Can you please rebase one last time? I swear this isn't normal. 😅

No worries 🙃 Done.

**Comment by emirot** [comment]
@pmalek  why partially ?

**Comment by pmalek** [comment]
@emirot Partially because IMHO there should be no need to use globals.

On top of that as I look at this now there are places where `globalSchema` is accessed without a lock: e.g. in https://github.com/kubernetes-sigs/kustomize/blob/ed09399cd1fa66b9290058ea909025385400f434/kyaml/openapi/openapi.go#L318

**Pull request merged**